packages/graphql-codegen-cli/src/utils/watcher.ts

```diff
@@ -136,32 +133,82 @@ export const createWatcher = (
 
     debugLog(`[Watcher] Started`);
 
-    const shutdown = () => {
+    const shutdown = (
+      /** Optional callback to execute after shutdown has completed its async tasks */
+      afterShutdown?: () => void
+    ) => {
       isShutdown = true;
       debugLog(`[Watcher] Shutting down`);
       log(`Shutting down watch...`);
-      watcherSubscription.unsubscribe();
-      lifecycleHooks(config.hooks).beforeDone();
+
+      const pendingUnsubscribe = watcherSubscription.unsubscribe();
+      const pendingBeforeDoneHook = lifecycleHooks(config.hooks).beforeDone();
+
+      if (afterShutdown && typeof afterShutdown === 'function') {
+        Promise.allSettled([pendingUnsubscribe, pendingBeforeDoneHook]).then(afterShutdown);
+      }
```

Comment:
This adds an optional callback to the `shutdown` handler for it to call after it's completed its async tasks. In normal usage, this callback is not supplied, which makes sense because `process.once()` event handlers are not supposed to wait for any asynchronous code. And in that case, the logic is unchanged: each async operation is queued immediately and there is no waiting for them to complete (the program will just exit).

But in tests, we want to be able to cancel the watcher, and gracefully shutdown. So in that case we give the `shutdown` handler a callback, and we wait for the two asynchronous events (`.unsubscribe()`, and `.beforeDone`) to complete before we call the callback, so that we ensure no async tasks are remaining on the queue when we use that callback to resolve the `pendingShutdown` promise outside of this loop.